src/devices/Common/System/Device/DelayHelper.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+
+namespace System.Device
+{
+    /// <summary>
+    /// Helpers for short waits.
+    /// </summary>
+    internal static class DelayHelper
+    {
+        // GetTimestamp() currently can take ~300ns. We hope to improve this to get better
+        // fidelity for very tight spins.
+        //
+        // SpinWait currently spins to approximately 1μs before it will yield the thread.
+
+        /// <summary>
+        /// Delay for at least the specified <paramref name="microseconds"/>.
+        /// </summary>
+        /// <param name="microseconds">The number of microseconds to delay.</param>
+        /// <param name="allowThreadYield">True to allow yielding the thread. On single-proc systems this will prevent all other code from running.</param>
+        public static void DelayMicroseconds(int microseconds, bool allowThreadYield)
+        {
+            long start = Stopwatch.GetTimestamp();
+            ulong elapsed = (ulong)(microseconds * Stopwatch.Frequency / 1_000_000);
+
+            if (!allowThreadYield)
+            {
+                do
+                {
+                    Thread.SpinWait(1);
+                } while (elapsed < (ulong)(Stopwatch.GetTimestamp() - start));
```

Comment:
Should this be `>` instead? The right side is actually the elapsed time since start and the left side is the delay duration. Similarly elsewhere.

Reply:
I think the check is correct but naming is confusing :) as @kouvel points out elapsed is the right side and left side is requested delay

Reply:
I think what we would want is `do ... while(elapsed < duration);`, since the condition is reversed above, the equivalent would then be `do ... while(duration > elapsed);`, isn't it?

Reply:
Yes, my terms are terrible. Updating to make it correct and hopefully clearer.

Reply:
@kouvel right :smile: